ws-server/src/server.rs

```diff
@@ -309,19 +315,32 @@ async fn background_task(
 	let stop_server2 = stop_server.clone();
 	let sink = MethodSink::new_with_limit(tx, max_response_body_size);
 
+	let mut submit_ping = tokio::time::interval(ping_interval);
+
 	middleware.on_connect();
 
 	// Send results back to the client.
 	tokio::spawn(async move {
 		while !stop_server2.shutdown_requested() {
-			if let Some(response) = rx.next().await {
-				// If websocket message send fail then terminate the connection.
-				if let Err(err) = send_ws_message(&mut sender, response).await {
-					tracing::warn!("WS send error: {}; terminate connection", err);
-					break;
+			select! {
+				rx_value = rx.next() => {
```

Comment:
this is cancel-safe?

Reply:
also I think you could do 

```
select! {
   Some(rx) = rx.next() => { }
   None = rx.next() => {}
   _ = submit_ping.tick() => {}
}
```

to get rid of that extra `else { break }`  branch.